### PR TITLE
🛂 Terraform state resource permissions

### PR DESCRIPTION
**Pull request by jacobwoffenden**
This pull request:

> [!WARNING]
> Before approving, please make sure you understand the context of the request and the consequences of it not working as expected

- Is part of https://github.com/ministryofjustice/analytical-platform/issues/5658
- Creates an S3 bucket policy as per [Terraform's guidance](https://developer.hashicorp.com/terraform/language/backend/s3#permissions-required)
  - Gives access to:
    - GitHub Actions' role
    - Analytical Platform engineer's role
    - Data Engineering engineer's roles
- Creates a DynamoDB resource policy as per [Terraform's guidance](https://developer.hashicorp.com/terraform/language/backend/s3#permissions-required)
  - Gives access to:
    - GitHub Actions' role
    - Analytical Platform engineer's role
    - Data Engineering engineer's roles
- Won't take effect until https://github.com/ministryofjustice/modernisation-platform/pull/8293 is applied

Signed-off-by: Jacob Woffenden <jacob.woffenden@justice.gov.uk>

### CODE REVIEW

File: terraform/aws/analytical-platform-management-production/terraform-state/data.tf [hunk 1]
```diff
@@ -7,3 +7,38 @@ data "aws_iam_session_context" "session" {
 
   arn = data.aws_caller_identity.session.arn
 }
+
+data "aws_region" "current" {}
+
+data "aws_iam_roles" "analytical_platform_team_access_role" {
```
Comment (julialawrence) [inline terraform/aws/analytical-platform-management-production/terraform-state/data.tf:13]: I understand why limiting access would be a good idea, but I think we might be called upon (maybe only initially) to help so worthwhile to add adminaccess roles in data-eng-prod unless you think it's better we're able to plan from management?

Comment (julialawrence) [reply terraform/aws/analytical-platform-management-production/terraform-state/data.tf:13]: And also data-prod and so on.

File: terraform/aws/analytical-platform-management-production/terraform-state/dynamodb-table.tf [hunk 2]
```diff
@@ -12,3 +12,94 @@ import {
...
+    actions = [
+      "dynamodb:DescribeTable",
+      "dynamodb:GetItem",
+      "dynamodb:PutItem",
+      "dynamodb:DeleteItem"
+    ]
+    resources = [module.state_locking.dynamodb_table_arn]
+    principals {
+      type        = "AWS"
+      identifiers = ["arn:aws:iam::${var.account_ids["analytical-platform-data-production"]}:role/aws-reserved/sso.amazonaws.com/${data.aws_region.current.name}/${one(data.aws_iam_roles.data_engineering_team_access_role_data_production_data_eng.names)}"]
```
Comment (julialawrence) [inline terraform/aws/analytical-platform-management-production/terraform-state/dynamodb-table.tf:97]: Yeah so data engineers will be able to plan from data eng account or data prod account. I want that too! Gimme that!!!

File: terraform/aws/analytical-platform-management-production/terraform-state/s3-bucket.tf [hunk 3]
```diff
@@ -1,3 +1,130 @@
...
+    actions = [
+      "s3:GetObject",
+      "s3:PutObject"
+    ]
+    resources = ["${module.state_bucket.s3_bucket_arn}/aws/analytical-platform-data-engineering-sandbox-a/*"]
+    principals {
+      type = "AWS"
+      identifiers = [
+        "arn:aws:iam::${var.account_ids["analytical-platform-data-engineering-sandbox-a"]}:role/aws-reserved/sso.amazonaws.com/${data.aws_region.current.name}/${one(data.aws_iam_roles.data_engineering_team_access_role_data_engineering_sandbox_a_admin.names)}",
+        "arn:aws:iam::${var.account_ids["analytical-platform-data-engineering-sandbox-a"]}:role/aws-reserved/sso.amazonaws.com/${data.aws_region.current.name}/${one(data.aws_iam_roles.data_engineering_team_access_role_data_engineering_sandbox_a_data_eng.names)}"
```
Comment (julialawrence) [inline terraform/aws/analytical-platform-management-production/terraform-state/s3-bucket.tf:98]: Just the same stuff as above, we want to add ourselves here.

File: terraform/aws/analytical-platform-management-production/terraform-state/data.tf [hunk 4]
```diff
@@ -7,3 +7,38 @@ data "aws_iam_session_context" "session" {
 
   arn = data.aws_caller_identity.session.arn
```
Comment (julialawrence) [inline terraform/aws/analytical-platform-management-production/terraform-state/data.tf:8]: Is the PR to change the assume role policy for the CI role coming up in another PR or are they getting a role of their own?

**Review by julialawrence — CHANGES_REQUESTED** [review]
Let's see if I understand this:

They get access to the data-eng-prod and data-eng-sandbox folders in the state bucket
They get access to a new dynamodb table for state locking (any reason we can't reuse the same one we use for ours?)
The change in githubactions role is trust policy in a different PR?

Looks great just a couple of changes/questions!

File: terraform/aws/analytical-platform-management-production/terraform-state/data.tf (hunk 4 above)
Comment (jacobwoffenden) [reply terraform/aws/analytical-platform-management-production/terraform-state/data.tf:8]: They are using our repository so AFAIK that doesn't need changing

File: terraform/aws/analytical-platform-management-production/terraform-state/data.tf (hunk 1 above)
Comment (jacobwoffenden) [reply terraform/aws/analytical-platform-management-production/terraform-state/data.tf:13]: I _thought_ I'd made it so **our** permissions and local workflow remain the same; as in

1. exec into management prod
1. init
1. ???
1. profit

DEs will assume the relevant DE role for that account

**Review by julialawrence — APPROVED** [review]
Approved

**Pull request merged**